ga4gh/gff3Parser.py

```diff
@@ -0,0 +1,321 @@
+"""
+Object representation of GFF3 data and parser for GFF3 files.
+
+See: http://www.sequenceontology.org/gff3.shtml
+"""
+
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import urllib
+import copy
+import re
+import gzip
+import bz2
+import collections
+
+GFF3_HEADER = "##gff-version 3"
+
+
+class GFF3Exception(Exception):
+    """
+    Exception associated with GFF3 data.
+    """
+    def __init__(self, message, fileName=None, lineNumber=None):
+        if fileName is not None:
+            message = "{}:{}: {}".format(fileName, lineNumber, message)
+        super(GFF3Exception, self).__init__(message)
+
+# characters forcing encode for columns
+_encodeColReStr = "\t|\n|\r|%|[\x00-\x1F]|\x7f"
+_encodeColRe = re.compile(_encodeColReStr)
+
+# characters forcing encode for attribute names or values
+_encodeAttrReStr = _encodeColReStr + "|;|=|&|,"
+_encodeAttrRe = re.compile(_encodeAttrReStr)
+
+
+def _encodeCol(v):
+    """
+    Encode a column if is has special characters.
+    """
+    if _encodeColRe.search(v):
+        return urllib.quote(v)
+    else:
+        return v
+
+
+def _encodeAttr(v):
+    """
+    Encode a attribute name or value if is has special characters.
+    """
+    if _encodeAttrRe.search(v):
+        return urllib.quote(v)
+    else:
+        return v
+
+
+class Feature(object):
+    """
+    One feature notation from a GFF3, missing attribute, as code by `.' in
+    the file are stored as None.
+    """
+
+    __slots__ = ("seqname", "source", "type", "start", "end", "score",
```

Comment:
Is there any other reason to use __slots__?  

Jerome Kelleher <notifications@github.com> writes:
> Why the slots? If it's a perf optimisation because we're making loads of
> Features we should have a quick one-line comment to say so.